### PR TITLE
feat(forge): Expand forge install error handling

### PR DESCRIPTION
**Pull request by lattejed**
## Motivation

`forge install` eats unknown errors and fails silently, reporting a successful install when there wasn't one.

## Solution

`install_as_submodule` now handles a (common?) error related to `git submodule add` failing when the added path is blocked by `.gitignore`.

In addition a catch-all was added to `install_as_submodule` and `install_as_folder` to report unexpected errors instead of failing silently.

The `--quiet` flag was added to prevent `git` from sending info messages to `stderr`, which is its default behavior.

### CODE REVIEW

**Review by gakonst — COMMENTED** [review]
good to see you back contributing :) i was under the impression that `--quiet` will supress all kinds of output, how does it help here?

File: cli/src/cmd/install.rs [hunk 1]
```diff
@@ -88,7 +88,8 @@ pub(crate) fn install(
 /// installs the dependency as an ordinary folder instead of a submodule
 fn install_as_folder(dep: &Dependency, libs: &Path) -> eyre::Result<()> {
     let output = Command::new("git")
-        .args(&["clone", &dep.url, &dep.name])
+        // --quiet prevents git from sending info messages to stderr
+        .args(&["clone", "--quiet", &dep.url, &dep.name])
```
Comment (mattsse) [inline cli/src/cmd/install.rs:92]: great!
but weird that the flag comes in between the subcommand and value, can we put this at the end of the array?

File: cli/src/cmd/install.rs [hunk 2]
```diff
@@ -124,7 +127,8 @@ fn install_as_folder(dep: &Dependency, libs: &Path) -> eyre::Result<()> {
 fn install_as_submodule(dep: &Dependency, libs: &Path, no_commit: bool) -> eyre::Result<()> {
     // install the dep
     let output = Command::new("git")
-        .args(&["submodule", "add", &dep.url, &dep.name])
+        // --quiet prevents git from sending info messages to stderr
+        .args(&["submodule", "add", "--quiet", &dep.url, &dep.name])
```
Comment (mattsse) [inline cli/src/cmd/install.rs:131]: same

File: cli/src/cmd/install.rs [hunk 3]
```diff
@@ -141,6 +145,16 @@ fn install_as_submodule(dep: &Dependency, libs: &Path, no_commit: bool) -> eyre:
         )
     } else if stderr.contains("not a git repository") {
         eyre::bail!("\"{}\" is not a git repository", &dep.url)
+    } else if stderr.contains("paths are ignored by one of your .gitignore files") {
+        let error = stderr
+            .trim()
+            .split('\n')
```
Comment (mattsse) [inline cli/src/cmd/install.rs:151]: I believe there's also `.lines()`

**Review by mattsse — CHANGES_REQUESTED** [review]
smol nits, but lgtm

File: cli/src/cmd/install.rs (hunk 1 above)
Comment (lattejed) [reply cli/src/cmd/install.rs:92]: Neither of these commands picked it up after the positional args. Checking `man git-clone` this is the correct position in this case.

File: cli/src/cmd/install.rs (hunk 2 above)
Comment (lattejed) [reply cli/src/cmd/install.rs:131]: In this case the correct position is apparently after `submodule` so I'll update this for consistency.

File: cli/src/cmd/install.rs (hunk 3 above)
Comment (lattejed) [reply cli/src/cmd/install.rs:151]: Good catch, thanks!

File: cli/src/cmd/install.rs (hunk 1 above)
Comment (lattejed) [reply cli/src/cmd/install.rs:92]: FYI, I'm going to remove these flags anyway.

**Comment by lattejed** [comment]
@gakonst `--quiet` in this case prevents non-error messages from heading to stderr. Status messages are sent to stderr in normal operation. The flag does not suppress errors.

I thought this was a `git` quirk but it looks like it's fairly common. The more correct way to deal with it is to check the exit code. I'll update it.

Thanks! I'm definitely around but have to prioritize other things at the moment.

**Review by gakonst — APPROVED** [review]
nice - lgtm!

**Pull request merged**